main/src/ca/uwaterloo/flix/language/phase/Weeder.scala

```diff
@@ -585,7 +585,136 @@ object Weeder extends Phase[ParsedAst.Program, WeededAst.Program] {
 
-        case ParsedAst.Expression.FNil(sp1, sp2) =>
+        case ParsedAst.Expression.ArrayLit(sp1, elms, sp2) =>
+          @@(elms.map(e => visit(e, unsafe))) map {
+            case es => WeededAst.Expression.ArrayLit(es, mkSL(sp1, sp2))
+          }
+
+        case ParsedAst.Expression.ArrayNew(sp1, elm, len, sp2) =>
+          @@(visit(elm, unsafe), visit(len, unsafe)) map {
+            case(es, ln) => WeededAst.Expression.ArrayNew(es, ln, mkSL(sp1, sp2))
+          }
+
+        case ParsedAst.Expression.ArrayLoad(exp1, exp2, sp2) =>
+          val sp1 = leftMostSourcePosition(exp1)
+          val loc = mkSL(sp1, sp2)
+
+          @@(visit(exp1, unsafe), visit(exp2, unsafe)) map {
+            case (ex1, ex2) => WeededAst.Expression.ArrayLoad(ex1, ex2, loc)
+          }
+
+        case ParsedAst.Expression.ArrayStore(exp1, exp2, exp3, sp2) =>
+          val sp1 = leftMostSourcePosition(exp1)
+          val loc = mkSL(sp1, sp2)
+
+          @@(visit(exp1, unsafe), visit(exp2, unsafe), visit(exp3, unsafe)) map {
+            case(ex1, ex2, ex3) => WeededAst.Expression.ArrayStore(ex1, ex2, ex3, loc)
+          }
+
+        case ParsedAst.Expression.ArrayLength(sp1, exp, sp2) =>
+          visit(exp, unsafe) map {
+            case es => WeededAst.Expression.ArrayLength(es, mkSL(sp1, sp2))
+          }
+
+        case ParsedAst.Expression.ArraySlice(exp1, exp2, exp3, sp2) =>
+          val sp1 = leftMostSourcePosition(exp1)
+          val loc = mkSL(sp1, sp2)
+
+          @@(visit(exp1, unsafe), visit(exp2, unsafe), visit(exp3, unsafe)) map {
+            case(ex1, ex2, ex3) => WeededAst.Expression.ArraySlice(ex1, ex2, ex3, loc)
+          }
+
+        case ParsedAst.Expression.ArraySliceNoEndIndex(exp1, exp2, sp2) =>
+          val sp1 = leftMostSourcePosition(exp1)
+          val loc = mkSL(sp1, sp2)
+
+          @@(visit(exp1, unsafe), visit(exp2, unsafe)) map {
+            case(ex1, ex2) => WeededAst.Expression.ArraySliceNoEndIndex(ex1, ex2, loc)
+          }
+
+        case ParsedAst.Expression.ArraySliceNoStartIndex(exp1, exp2, sp2) =>
+          val sp1 = leftMostSourcePosition(exp1)
+          val loc = mkSL(sp1, sp2)
+
+          @@(visit(exp1, unsafe), visit(exp2, unsafe)) map {
+            case(ex1, ex2) => WeededAst.Expression.ArraySliceNoStartIndex(ex1, ex2, loc)
+          }
+
+        case ParsedAst.Expression.VectorLit(sp1, elms, sp2) =>
+          /*
+           * Rewrites empty vectors to Unit.
+           */
+          @@(elms.map(e => visit(e, unsafe))) map{
+            case Nil =>
+              val loc = mkSL(sp1, sp2)
+              WeededAst.Expression.Unit(loc)
+            case es => WeededAst.Expression.VectorLit(es, mkSL(sp1, sp2))
+          }
+
+        case ParsedAst.Expression.VectorNew(sp1, elm, len, sp2) =>
+          visit(elm, unsafe) flatMap {
+            case e => len match {
+              case ParsedAst.Literal.Int32(sp1, sign, digits, sp2) => toInt32(sign, digits, mkSL(sp1, sp2)) flatMap {
```

Comment:
This code can be simplified if you change the type of VectorNew to have an Int32 literal as its type for len.